tests/TestCase/ORM/EntityTest.php

```diff
@@ -46,6 +46,31 @@ public function testSetOneParamNoSetters()
         $this->assertSame(1, $entity->id);
         $this->assertEquals(1, $entity->getOriginal('id'));
         $this->assertEquals('bar', $entity->getOriginal('foo'));
+
+        $entity->set('level1.level2.level3', 'three');
+        $this->assertSame('three', $entity->level1->level2->level3);
+        $this->assertEquals('three', $entity->getOriginal('level1.level2.level3'));
+        $entity->set('level1.level2.level3', '333');
+        $this->assertEquals('three', $entity->getOriginal('level1.level2.level3'));
```

Comment:
> Totally agree, but the question is how to stop when we run out of Entities (but could theoretically descend into an array that lacks the ability to track dirty state.) 

In the EntityContext code I did [some sniffing](https://github.com/cakephp/cakephp/blob/master/src/View/Form/EntityContext.php#L268-L338) and poking to find the leaf entity object.